### ISSUE TITLE
[CS2] Polyfills?

### ISSUE THREAD

**Issue by GeoffreyBooth** [jashkenas/coffeescript#4674]
(From @helixbass https://github.com/jashkenas/coffeescript/pull/4652#issuecomment-326139705):

Babel outputs polyfills for code _it_ generates, but not for code the CoffeeScript compiler outputs. In other words, sending Babel this:
```js
var obj2 = {...obj1};
```
will result in [this](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&circleciRepo=&code_lz=G4QwTgBA9gRgVgJggXggbwHRdnAjAXwG4g&debug=false&evaluate=false&lineWrap=true&presets=stage-3%2Cenv&prettier=false&showSidebar=true&targets=&version=6.26.0):
```js
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var obj2 = _extends({}, obj1);
```
But sending it the CoffeeScript compiler output of `obj2 = {obj1...}`:
```js
var obj2;

obj2 = Object.assign({}, obj1);
```
Babel leaves it unmodified. Babel’s [babel-preset-env](https://github.com/babel/babel-preset-env) converts _syntax_ for compatibility with browser presets, but it doesn’t add polyfills that your target browsers may need (such as for `Object.assign` in this case).

I for one was under the mistaken impression that Babel covered us in such cases. I’ve done some research and there doesn’t appear to be any tool out there to scan your code to identify which polyfills you’ll need. There’s [some discussion](https://github.com/babel/babel-preset-env/issues/294) that Babel should do this, but it seems a ways off. Here’s [a great article](https://philipwalton.com/articles/loading-polyfills-only-when-needed/#areas-for-improvement) that mentions the same idea. There’s also http://jscc.info/, but it appears to be several years out of date (it doesn’t know about `Object.assign`, for example).

In #4526 we removed polyfills for [`Array.indexOf`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf) and [`Function.prototype.bind`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind), which are supported in Internet Explorer 9+. In #4493 we added [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) (not supported in any version of Internet Explorer, even 11) without a polyfill.

As far as I can tell, reading through `nodes.coffee`, those are the only browser native functions that might possibly need polyfilling. The others we use—`{}.hasOwnProperty`, `[].slice`, `[].splice`—are so old that MDN just lists “(yes)” instead of a browser version when they were first supported.

We should do one of the following:

1. Update the docs to mention that polyfills are required if Internet Explorer support is desired (spelling out which polyfills for which versions based on the above).
1. Polyfill `Object.assign` similarly to how we used to polyfill `Array.indexOf`, probably using the same polyfill that Babel outputs (see second code block above). Mention in the docs that if support for IE8 is desired, the user should polyfill `Array.indexOf` and `Function.prototype.bind`.
1. Polyfill all three, and leave the docs alone. We would write the polyfills like Babel does, so that the native method is used if available: `Object.assign || ...`.

**Comment by helixbass** [comment]
@GeoffreyBooth thanks for the useful summary. I think # 2 is the most sensible option. Haven't read the original discussion of removing `indexOf` and `bind` polyfills but targeting IE9+ (with, as you suggest, a mention in the docs) seems like the right choice. For `Object.assign`, you'd be forcing most people to complicate their build chain/code to include the polyfill so I don't think it's a good idea not to "polyfill" it ourselves

Opened #4675 which replaces the direct call to `Object.assign()` when compiling object spreads into a call to a new `_extends` utility (copied directly from Babel's)

**Comment by GeoffreyBooth** [comment]
Agreed. I added the docs note into your PR.

**Comment by GeoffreyBooth** [comment]
Fixed via #4675.
